dotenv-linter/src/schema.rs

```diff
@@ -0,0 +1,335 @@
+use anyhow::Result;
+use serde::Deserialize;
+use std::{collections::HashMap, fs::File, io::BufReader, path::Path};
+
+#[derive(Deserialize, Default)]
+#[serde(default)]
+struct ReadDotEnvSchema {
+    pub version: String,
+    pub allow_other_keys: bool,
+    pub entries: Vec<SchemaEntry>,
+}
+
+pub struct DotEnvSchema {
+    pub version: String,
+    pub allow_other_keys: bool,
+    pub entries: HashMap<String, SchemaEntry>,
+}
+#[derive(Deserialize, Default)]
+#[serde(default)]
+pub struct SchemaEntry {
+    pub key: String,
+    pub required: bool,
+    #[serde(rename = "type")]
+    pub value_type: SchemaValueType,
+    pub regex: Option<String>,
+}
```

Comment:
Since we do conversion, what do you think of using `enums`? It also would introduce some validation on schema reading

I mean something like:
```rust
enum EnvVariableKind {
  String { regex },
  Email,
  Nunber { min, max },
  ..
}
```

Reply:
I wondered about this too, I dont think so though. The only useful one at the moment is `String(Option<Regex>)` . If the idea is to encapsulate the entire definition of the 'rule' in the enum they get too complex. Imagine rules that say
- either KEY1 or KEY2 must be present but not both
- if KEY1 present then KEY2 must also be present
- these are the allowed values for KEY1 (could be a regex)
- ...

The code then still needs to look at other things in the schema definition , not just the type enum so my vote is no.

Reply:
If we need type-independent checks, it could be more extended wrapper, like:

```rs

entries: HashMap<String, EnvValidationRules>

struct EnvValidationRules {
   ty: EnvVariableType,
   excludes: Vec<String>, // keys which can't present in case of this one
   ... // common rules which doesn't depend on type
}

The main point here that there are inappropriate configs for particular types (i.e depends only on type). 
For now only `Regex` for string, but we already have understanding of validations for numbers and can have more.

It's kinda a good way to ensure schema is passed in correct way, and let's say there is no such confusing definitions:

```
        {
            "key": "SECRET",
            "required": true,
            "type": "Number",
            "regex": "^[a-z]*$"
        },
```